### ISSUE TITLE
Questions about customized DQN training

### ISSUE THREAD

**Issue by Timerunning** [aai-institute/tianshou#617]
Hello, I'd like to train a robust DQN based on tianshou. I need to extract replay data from buffer to compute the robust loss and standard loss, and then pass them to DQN network. I look at the documentation but found nothing, therefore asking how to implement that idea. Thanks!

**Comment by Trinkle23897** [comment]
Could you please state the detail of the idea?

**Comment by Timerunning** [comment]
Here is the process:

1. the main logic:

```python
data = replay_buffer.sample(args.batch_size)
if args.robust:
    loss, standard_loss, worst_case_loss = _compute_robust_loss(current_model, target_model, data, attack_epsilon,
                                                         kappa, args.gamma, device, args)
optimizer.zero_grad()
loss.backward()
optimizer.step()
```

2. _compute_robust_loss():

```python
def _compute_robust_loss(curr_model, target_model, data, epsilon, kappa, gamma, device, args):
    state, action, reward, next_state, done = data
    
    q_values      = curr_model(state)
    next_q_values = curr_model(next_state)
    next_q_state_values = target_model(next_state) 

    q_value       = q_values.gather(1, action.unsqueeze(1)).squeeze(1)
    next_q_value = next_q_state_values.gather(1, torch.argmax(next_q_values, 1, keepdim=True)).squeeze(1)
    expected_q_value = reward + gamma * next_q_value * (1 - done)
    
    standard_loss = torch.min((q_value - expected_q_value.detach()).pow(2), torch.abs(q_value - expected_q_value.detach()))

    #epsilon is attack distence
    upper, lower = network_bounds(curr_model.model, state, epsilon)
    onehot_labels = torch.zeros(upper.shape).to(device)
    onehot_labels[range(state.shape[0]), action] = 1
    
    if args.worse_bound:
        upper_diff = upper - q_values*(1-onehot_labels) - expected_q_value.detach().unsqueeze(1)*onehot_labels
        lower_diff = lower - q_values*(1-onehot_labels) - expected_q_value.detach().unsqueeze(1)*onehot_labels
        wc_diff = torch.max(torch.abs(upper_diff), torch.abs(lower_diff))
    else:
        worst_case = onehot_labels*lower+(1-onehot_labels)*upper
        wc_diff = torch.abs(worst_case - q_values*(1-onehot_labels) - expected_q_value.detach().unsqueeze(1)*onehot_labels)
    
    #sum over output layer, mean only in batch dimension
    worst_case_loss = torch.sum(torch.min(wc_diff.pow(2), wc_diff), dim=1).mean()
    
    standard_loss = standard_loss.mean()
    
    loss = (kappa*(standard_loss)+(1-kappa)*(worst_case_loss))
    
    return loss, standard_loss, worst_case_loss
```

**Comment by Trinkle23897** [comment]
Just create a new class `RobustDQNPolicy` that inherits from `DQNPolicy`, and then overwrite the `learn` function.
This is the original DQN learn function:
https://github.com/thu-ml/tianshou/blob/7f23748347d6bf4aebce3931f7e57291012cd98d/tianshou/policy/modelfree/dqn.py#L164-L178

**Comment by Timerunning** [comment]
Thanks for your suggestion! I have a follow-up question: `batch.weight` is not present in paper "Human-level control through deep reinforcement learning". I don't quite understand what `weight` is for, and should I replace the code as `batch.weight = td_error+worst_case_loss`?

**Comment by Jimenius** [comment]
`batch.weight` is related to Prioritized Experience Replay. If you don't use `PrioritizedReplayBuffer`, just ignore it.
